src/providers/indexeddb.js

```diff
@@ -10,67 +10,90 @@ var indexedDB = global.indexedDB       ||
                 global.webkitIndexedDB ||
                 global.msIndexedDB;
 
+function TransactionBuilder(db, mode) {
+  this.db = db;
+  this.mode = mode;
+}
+TransactionBuilder.prototype.build = function() {
+  // Inspired by https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-indexeddb/src/util.js#L39-L49
+  try {
+    this.transaction = this.db.transaction(FILE_STORE_NAME, this.mode);
+  } catch (err) {
+    this.error = err;
+  }
+};
+
+
 function IndexedDBContext(db, mode) {
-  var transaction = db.transaction(FILE_STORE_NAME, mode);
-  this.objectStore = transaction.objectStore(FILE_STORE_NAME);
+  this.db = db;
+  this.mode = mode;
 }
 
 IndexedDBContext.prototype.clear = function(callback) {
-  try {
-    var request = this.objectStore.clear();
-    request.onsuccess = function(event) {
-      callback();
-    };
-    request.onerror = function(error) {
-      callback(error);
-    };
-  } catch(e) {
-    callback(e);
+  var builder = new TransactionBuilder(this.db, this.mode);
+  builder.build();
+  if(builder.error) {
+    return callback(builder.error);
   }
+
+  var transaction = builder.transaction;
+  var objectStore = transaction.objectStore(FILE_STORE_NAME);
```

Comment:
Indeed we should, confirmed here https://twitter.com/asutherland/status/866498053519192064.  I spent a long time reading other IDB wrapper libraries, and they are all over the map in terms of what they do with `InvalidStateError` handling--most do nothing.  I'll fix.